src/coreclr/jit/valuenum.cpp

```diff
@@ -12560,17 +12575,52 @@ void Compiler::fgValueNumberCall(GenTreeCall* call)
     }
     else
     {
+        bool handled = false;
         if (call->TypeGet() == TYP_VOID)
         {
             call->gtVNPair.SetBoth(ValueNumStore::VNForVoid());
         }
         else
         {
-            call->gtVNPair.SetBoth(vnStore->VNForExpr(compCurBB, call->TypeGet()));
+            // Optimize Type.GetTypeFromHandle(TypeHandleToRuntimeTypeHandle(clsHandle)) to a frozen handle.
+            if (call->IsSpecialIntrinsic(this, NI_System_Type_GetTypeFromHandle))
+            {
+                VNFuncApp bitcastFuncApp;
+                VNFuncApp typeHandleFuncApp;
+                ValueNum  argVN = call->gtArgs.GetArgByIndex(0)->GetNode()->gtVNPair.GetConservative();
+                if (vnStore->GetVNFunc(argVN, &bitcastFuncApp) && (bitcastFuncApp.m_func == VNF_BitCast))
```

Comment:
Can we move this into its own method? It is a good pattern to do specific folds like this outside the top-level methods, to avoid them growing in size.

This also needs an assert that we'll only see `NI_System_Type_GetTypeFromHandle` on NAOT.